examples/evaluate_ensemble_judge.py

```diff
@@ -0,0 +1,124 @@
+import json
+
+import yaml
+from unitxt import get_logger
+from unitxt.api import evaluate, load_dataset
+from unitxt.blocks import Task, TaskCard
+from unitxt.inference import (
+    IbmGenAiInferenceEngine,
+    IbmGenAiInferenceEngineParams,
+)
+from unitxt.llm_as_judge import LLMAsJudge
+from unitxt.loaders import LoadFromDictionary
+from unitxt.metrics import (
+    LogisticRegressionMetricsEnsemble,
+    RandomForestMetricsEnsemble,
+)
+from unitxt.templates import InputOutputTemplate, TemplatesDict
+from unitxt.text_utils import print_dict
+
+logger = get_logger()
+
+platform = "ibm_gen_ai"
+gen_params = IbmGenAiInferenceEngineParams(max_new_tokens=256)
+
+data = {
+    "test": [
+        {"conversation": "user: is eugene from walking dead really a scientist"},
+        {"conversation": "user: is eugene from walking dead really a scientist"},
+        {"conversation": "user: is eugene from walking dead really a scientist"},
+    ]
+}
+
+predictions = [
+    "In the Walking Dead series, Eugene Porter initially claims to be a scientist who knows the cure to the zombie plague. However, it is eventually revealed that he is not a scientist, but a high school science teacher who lied about knowing the cure to manipulate other survivors into helping him.",
+    "No, Eugene Porter, played by Josh McDermitt, is not a scientist in real life. In the Walking Dead TV series, he portrays a character who initially appears to be a scientist with knowledge of a cure for the zombie outbreak. However, it is later revealed that he is actually a high school science teacher who lied about his credentials to manipulate other survivors into taking him to Washington D.C.",
+    "thank you for the question.",
+]
+
+config_filepath = "prepare/templates/response_assessment/judges/config_judges.json"
```

Comment:
Since there needs to be alignment between the individual LLM as judges  and weights.  I would think that the Ens

In the future, I would think it

ensemble_metric = RandomForestMetricsEnsemble(metric_list)
ensemble_metric.train(dataset) # would perform random forest model training of a dataset with input text and expected rating, then store the results in the weights field as json string.
add_to_catalog(ensemble_metric."metrics.llm_as_judge.relevance.ensemble_random_forest_4_ibmgenai_judges")

Reply:
Right now , I think it can be 

ensemble_metric = RandomForestMetricsEnsemble(metric_list)
ensemble_metric.load_weights("prepare/templates/response_assessment/judges/ensemble_relevance_v1.json")
add_to_catalog(ensemble_metric."metrics.llm_as_judge.relevance.ensemble_random_forest_4_ibmgenai_judges")

Reply:
Now the use of the metric can just 
    loader=LoadFromDictionary(data=data),
    task=Task(
        input_fields={"conversation": "str"},
        reference_fields={},
        prediction_type="str",
        metrics=["metrics.llm_as_judge.relevance.ensemble_random_forest_4_ibmgenai_judges"],
    ),

Reply:
Assuming load_weights is a function inside `RandomForestMetricsEnsemble` 
```
class RandomForestMetricsEnsemble(MetricsEnsemble):
    ensemble_model = None
    def prepare(self):
        super().prepare()
    def load_weights(self, json_file):
        with open(json_file, 'r') as file:
            json_data = json.load(file)
            
        self.ensemble_model = self.decode_forest(json_data)
```

How does evaluate method in `evaluated_dataset = evaluate(predictions=predictions, data=test_dataset)` find these weights?

The following code throws me this error:
```
ensemble_metric = RandomForestMetricsEnsemble(metrics=metric_lst)
ensemble_metric.load_weights("prepare/templates/response_assessment/judges/ensemble_relevance_v1.json")
add_to_catalog(ensemble_metric, "metrics.llm_as_judge.relevance.ensemble_random_forest_v1_ibmgenai_judges", overwrite=True)

card = TaskCard(
    loader=LoadFromDictionary(data=data),
    task=Task(
        input_fields={"conversation": "str"},
        reference_fields={},
        prediction_type="str",
        metrics = ["metrics.llm_as_judge.relevance.ensemble_random_forest_v1_ibmgenai_judges"],
    ),
    templates=TemplatesDict(
        {
            "simple": InputOutputTemplate(
                input_format="{conversation}",
                output_format="",
            )
        }
    ),
)
test_dataset = load_dataset(card=card, template_card_index="simple")["test"]
evaluated_dataset = evaluate(predictions=predictions, data=test_dataset)
```
```
File ~/Desktop/chat/unitext3/src/unitxt/api.py:91, in evaluate(predictions, data)
     [90](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/api.py:90) def evaluate(predictions, data) -> List[Dict[str, Any]]:
---> [91](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/api.py:91)     return _compute(predictions=predictions, references=data)

File ~/Desktop/chat/unitext3/src/unitxt/metric_utils.py:264, in _compute(predictions, references, flatten, split_name, calc_confidence_intervals)
    [261](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/metric_utils.py:261)     multi_stream = operator(multi_stream)
    [263](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/metric_utils.py:263) stream = multi_stream[split_name]
--> [264](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/metric_utils.py:264) return list(stream)

File ~/Desktop/chat/unitext3/src/unitxt/generator_utils.py:16, in ReusableGenerator.__iter__(self)
     [15](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/generator_utils.py:15) def __iter__(self):
---> [16](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/generator_utils.py:16)     yield from self.activate()

File ~/Desktop/chat/unitext3/src/unitxt/operator.py:366, in InstanceOperator._process_stream(self, stream, stream_name)
    [364](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/operator.py:364) except Exception as e:
    [365](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/operator.py:365)     if _index is None:
--> [366](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/operator.py:366)         raise e
    [367](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/operator.py:367)     else:
    [368](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/operator.py:368)         raise ValueError(
    [369](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/operator.py:369)             f"Error processing instance '{_index}' from stream '{stream_name}' in {self.__class__.__name__} due to: {e}"
    [370](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/operator.py:370)         ) from e

File ~/Desktop/chat/unitext3/src/unitxt/operator.py:362, in InstanceOperator._process_stream(self, stream, stream_name)
...
   [4764](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/metrics.py:4764) print("Ensemble model in ensemble()", self.ensemble_model)
-> [4765](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/metrics.py:4765) score = self.ensemble_model.predict([prediction_lst])
   [4766](https://file+.vscode-resource.vscode-cdn.net/Users/vrajshah/Desktop/chat/unitext3/examples/~/Desktop/chat/unitext3/src/unitxt/metrics.py:4766) return score.tolist()[0]

AttributeError: 'NoneType' object has no attribute 'predict'
```

Reply:
I think the idea is that 'load_weights()' will only store the weights.  The process() will lazy create the model.

class RandomForestMetricsEnsemble(MetricsEnsemble):
    ensemble_model = None
    def prepare(self):
        super().prepare()
    def load_weights(self, json_file):
        with open(json_file, 'r') as file:
            self.weights = json.load(file)
   
    def ensemble(self, instance):
        if ( self.ensemble_model is None)
             self.ensemble_model = self.decode_forest(self.weights)
        print("Accessing ensemble_model in ensemble():", self.ensemble_model)
        prediction_lst = []
        for i, metric in enumerate(self.metrics):
            prediction_lst.append(
                instance["score"]["instance"][
                    self.get_prefix_name(i) + metric.main_score
                ]
            )
        print("Ensemble model in ensemble()", self.ensemble_model)
        score = self.ensemble_model.predict([prediction_lst])
        return score.tolist()[0]

Reply:
Then, when we save the model 

ensemble_metric.load_weights("prepare/templates/response_assessment/judges/ensemble_relevance_v1.json")
add_to_catalog(ensemble_metric."metrics.llm_as_judge.relevance.ensemble_random_forest_4_ibmgenai_judges")

 The weights will be store in the catalog, can could be reused when the metric is loaded again in a separate run.

Reply:
When I run the above code, it doesn't store the weights in the catalog. For instance, the below test code would only store {"weights": "test", ... } in the catalog, but wouldn't update the weights attribute w/ "test_load_weights". The same happens when `MetricsEnsemble` is used instead of `RandomForestMetricsEnsemble.` can you pls confirm this?

```
ensemble_metric = RandomForestMetricsEnsemble(metrics=metric_lst, weights = "test")
ensemble_metric.load_weights("../prepare/templates/response_assessment/judges/ensemble_relevance_v1.json")
add_to_catalog(ensemble_metric
               , "metrics.llm_as_judge.relevance.ensemble_random_forest_4_ibmgenai_judges"
               , overwrite=True
               )
```

class RandomForestMetricsEnsemble(MetricsEnsemble):
  def load_weights(self, json_file):
          self.weights = "test_load_weights"

Reply:
Ok. I thought it would work out of the box - but we need to look into that (I'll need @elronbandel's input).  
As a temporary solution , you can make the load_weights a static method and wt the weights in the initialization.

```
weights = RandomForestMetricsEnsemble.load_weights("../prepare/templates/response_assessment/judges/ensemble_relevance_v1.json")
ensemble_metric = RandomForestMetricsEnsemble(metrics=metric_lst, weights = weights)
add_to_catalog(ensemble_metric
               , "metrics.llm_as_judge.relevance.ensemble_random_forest_4_ibmgenai_judges"
               , overwrite=True
               )
```

Reply:
Thank you, that fixed the issue.